### PR TITLE
add randomly generated environment

### LINKED ISSUE #55: Create an Infinite, scrolling background
Create a background that
- [x] scrolls with the camera in any direction infinitely
- [ ] has terrain elements in a non-repeating pattern

### PR DESCRIPTION
**Pull request by loteque**
- [x] Environment can be scrolled infinitely in any direction. 
  - ~~Currently the terrain chunk uses the random number generator to place terrain elements, while the environment only uses one chunk. This leads to a repeating pattern as the player scrolls through the level.~~
- [x] Make terrain elements in a non-repeating pattern.

closes #55

### CODE REVIEW

**Comment by jonathandavidlewis** [comment]
Are you able to add videos or gifs to the PRs?

I recorded one for this.

https://user-images.githubusercontent.com/23508546/220175337-a4072082-74be-4b4b-95ad-1b202de33ce3.mp4

File: project/src/TerrainSprites.gd [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+	return tex
+
+func get_sprite_size(sprite: Sprite) -> Vector2:
+	var sprite_size: Vector2
+	sprite_size = Vector2(sprite.texture.get_width(), sprite.texture.get_height())
+
+	return sprite_size
+
+func generate_random_number(minimum: int, maximum: int) -> int:
+	var rand: int
```
Comment (jonathandavidlewis) [inline project/src/TerrainSprites.gd:50]: rand is initialized in the same scope as this, so no need to separate declaration and initialization.

File: project/src/TerrainSprites.gd [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+	var rng = RandomNumberGenerator.new()
+	rng.randomize()
```
Comment (jonathandavidlewis) [inline project/src/TerrainSprites.gd:52]: Can we pull out the call to `randomize` into something called once per game session? As I understand it, that is more typical.

File: project/src/TerrainSprites.gd [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+	
+	return tex
+
+func get_sprite_size(sprite: Sprite) -> Vector2:
+	var sprite_size: Vector2
+	sprite_size = Vector2(sprite.texture.get_width(), sprite.texture.get_height())
+
+	return sprite_size
+
+func generate_random_number(minimum: int, maximum: int) -> int:
```
Comment (jonathandavidlewis) [inline project/src/TerrainSprites.gd:49]: How do you feel about pulling out random number generation into a separate file so we can pull it into other files from there? We could do this after the merge, just curious about your thoughts. I was making something also and started a random number script.

File: project/src/TerrainSprites.gd [hunk 4]
```diff
@@ -0,0 +1,55 @@
+tool
+extends Node2D
+
+export var terrain_textures: Array
+export var NUM_TERRAIN_SPRITES: int = 15
+
+const SCREEN_SPACE = Vector2(512, 448)
+const HALF_SCREEN_SPACE = Vector2(SCREEN_SPACE.x / 2, SCREEN_SPACE.y / 2)
+const QUART_SCREEN_SPACE = Vector2(SCREEN_SPACE.x / 4, SCREEN_SPACE.y / 4)
+const CHUNK_SIZE = Vector2(1024, 896)
```
Comment (jonathandavidlewis) [inline project/src/TerrainSprites.gd:10]: Can this be derived from `SCREEN_SPACE`? Or is it decoupled?

**Review by jonathandavidlewis — CHANGES_REQUESTED** [review]
This is a great effect!!! So stoked! Do you have time today to chat about the suggested changes?

File: project/src/CameraController.gd [hunk 5]
```diff
@@ -13,4 +16,9 @@ func move(delta) -> void:
 	if Input.is_action_pressed("ui_left"):
 		position.x -= speed * delta
 	if Input.is_action_pressed("ui_right"):
-		position.x += speed * delta
\ No newline at end of file
+		position.x += speed * delta
+
+func signal_cam_pos_updated() -> void:
+	frame = ++frame
```
Comment (jonathandavidlewis) [inline project/src/CameraController.gd:22]: ++ doesn't pre-increment in gdscript. I learned there's not a post-increment either. As far as I can tell, we would have to use += or = and +.  Therefore, this number is always 0. I also can't figure out what ++ does and why it does not throw an error. Let me know if you figure this out.

Comment (loteque) [reply project/src/CameraController.gd:22]: Omg that is wild.  I'll look into it.

File: project/src/BackgroundGenerator.gd [hunk 6]
```diff
@@ -0,0 +1,38 @@
+extends Node2D
+
+onready var grid: TileMap = get_node("TileMap")
+
+var tile_background: Resource = load("res://src/BackgroundChunk.tscn")
+var camera_pos: Vector2
```
[the remark this answers is not in the document]
Comment (loteque) [reply project/src/BackgroundGenerator.gd:6]: I agree with this change. good catch!

File: project/src/CameraController.gd [hunk 7]
```diff
@@ -2,7 +2,10 @@ extends Position2D
```
[the remark this answers is not in the document]
Comment (loteque) [reply project/src/CameraController.gd:2]: Yeah let's merge the camera with the player controller in a later issue.

File: project/src/CameraController.gd [hunk 8]
```diff
@@ -2,7 +2,10 @@ extends Position2D
 
 export(float) var speed = 10.5
 
+var frame: int
```
[the remark this answers is not in the document]
Comment (loteque) [reply project/src/CameraController.gd:5]: yeah let's do that!

File: project/src/TerrainSprites.gd (hunk 4 above)
Comment (loteque) [reply project/src/TerrainSprites.gd:10]: It is decoupled.

File: project/src/TerrainSprites.gd (hunk 3 above)
Comment (loteque) [reply project/src/TerrainSprites.gd:49]: yeah. I'll hand this off to. thanks!

**Review by jonathandavidlewis — APPROVED** [review]

**Pull request merged**